spec/src/main/asciidoc/config.asciidoc

```diff
@@ -0,0 +1,147 @@
+//
+// Copyright (c) 2019 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+[[config]]
+== Common Configuration Patterns
+
+Many MicroProfile specifications make use of the Configuration specification for
+setting or overriding behaviour characteristics. Listed below are a list of recommended usage
+patterns that are already adopted by one or more of the MicroProfile specifications.
+
+[[patterns]] 
+=== Patterns
+
+[[propertyPrefix]]
+==== Configuration Property Name Prefixes
+
+In order to prevent collision it is recommended that properties that are not
+named to match application class names are prefixed with "mp." followed by the specification's short name:  `mp.<spec-name>`.
+
+----
+// For example, the Messaging Specification has an @Outgoing annotation:
+   mp.messaging.outgoing.attributename=3
+----
+
+[[qualifyingClassNames]]
+==== Properties That Relate To Application Classes
+
+All references to users' application classes in property names should be fully package qualified.
+
+There are places where an annotation class name appears or a specification sourced class name (e.g. Context Propagation) appears in configuration property names, these class names are not package qualified. Should these start with `mp.<spec-name>`? In this case `mp.ft`?
+----
+// For example: 
+   Retry/maxRetries=30
+//or should this be?
+   mp.ft.retry.maxRetries=30
+----
+This is probably the right choice from a usability point of view. 
+This is particularly not a problem where the scope of the configuration property identifies a particular Java file (which will have resolved the package name via an import).
+It could become more of a problem for service wide defaults if more than one MicroProfile service
+has a class with the same name. 
+The result is that we should have exclusive class names in MicroProfile for classes that might need configuration override in the process environment. 
+The 'first' specification that owns a class name should own that name exclusively. 
+
+[[fieldAnnotations]]
+==== Properties That Relate To Annotated Fields
```

Comment:
For Fields and Methods I think it makes sense to have some kind of short format.

The current naming is very verbose and prone to typos because of it.

Could the Annotation define a `key` for configuration in itself? Which can be overridden by the developer if they chose, but there's a sensible non clashing default `key`?

Reply:
I quite like this idea @kenfinnigan. Perhaps this would benefit from a  'first' spec working
group to introduce this pattern and iron it out with a use case rather then promote the usage pattern than us proposing something with no customer #1 in mind. Sort of similarly, in the MicroProfile Reactive Messaging spec we have the notion of a 'Connector' name which is used as a parameter in a CDI qualifier. This connector name label can act as an identifier for connector propeties that are applied to any Incoming/Outgoing annotated methods that refer to a channel that uses that connector. (OK only sort of similar :-) )

Reply:
Any thoughts on the best spec to try this out with?

Reply:
Ken - I am not au-fait with the current state of them all - what could be at the right stage and which group might like the idea. One thought - if the customer pain is a typo in the config  - perhaps we should spec Deployment errors if an attribute has a mp.<spec> prefix but the matching spec's CDi extension can't find any effect of the property?

Reply:
Fault Tolerance could be applicable or Context Propogation...best discussed in an issue in their repos?

Reply:
@kenfinnigan You are welcome to add a short format as an additional subsequent feature. I can't realistically resolve it in this PR. Fair idea though.